src/io.c

```diff
@@ -143,15 +143,35 @@ unsigned io_fprintln(const char *fname, const char *fmt, ...)
  */
 void io_init(const char *cfile, const char *datadir, const char *confdir)
 {
-	if (!datadir) {
-		if (!(datadir = getenv("HOME")))
-			datadir = ".";
-		asprintf(&path_ddir, "%s%s", datadir, "/" DIR_NAME);
-	} else
+	char* home_dir = getenv("HOME");
+
+	if (!datadir)
+		if (home_dir) {
+			asprintf(&path_ddir, "%s%s", home_dir, "/" DIR_NAME_LEGACY);
+			if (!io_dir_exists(path_ddir)) {
+				if ((path_ddir = getenv("XDG_DATA_HOME")))
+					asprintf(&path_ddir, "%s%s", path_ddir, "/" DIR_NAME);
```

Comment:
I think this creates a memory since you do not free the memory you allocated with the previous asprintf() call.